### DISCUSSION TITLE
[Question] I2C module in RPI

### DISCUSSION THREAD

**Discussion by ayoubSoussi** [modm-io/modm#704] [Q&A]
Hi,

I am trying to add a new I2C module to the Raspberry Pi Platform. It's basically a wrapper of the library WiringPi.
In the file I2C_master.cpp, I'm a stuck in the implementation of a function that executes a transaction. I couldn't understand well how I2C_transaction s work exactly (for example what to do when the next transaction is a Restart).

This is the function that executes a transaction (called in the startTransaction function)
![prob](https://user-images.githubusercontent.com/34512229/101986510-72901580-3c8e-11eb-919c-f9f32b566900.JPG)

I added just Write operation (as a first step).
To sum up, my question is 

- What should I do when the nextOperation is a Restart ?

Thanks.

**Comment by salkinium** [comment]
It should trigger the Repeated Start Condition on the bus: https://www.i2c-bus.org/repeated-start-condition
I'm not sure if the WiringPi API has a function for that, but this is how we encode write-read transactions as "write N bytes, restart, read M bytes".

**Comment by ayoubSoussi** [comment]
Apparently WiringPiI2C contains only basic functions for reading and writing, perhaps it would be better to interact with the i2c kernel driver directly.

**Comment by salkinium** [comment]
Perhaps? WiringPi seems to just wrap [`i2c_smbus_access`](https://github.com/WiringPi/WiringPi/blob/master/wiringPi/wiringPiI2C.c), but it's not clear to me if that API supports repeated starts either?

Looking through the search results for "RPi I2C" there seems to be a lot of issues and confusion about the I2C in general… 😬

**Comment by salkinium** [comment]
I mean you can try to do a write-stop-start-read sequence to get you started with something, but it's functionally different than a write-restart-read sequence and some devices will not work correctly. But it may work for your devices for now and I'll be happy to merge this with a big disclaimer about restart conditions that someone else can then fix.

**Comment by ayoubSoussi** [comment]
I want to use for an I2C SSD1306 LCD, and after looking at the SSD1306 driver I see that it uses restart so I'm not sure if it will work without modifying the SSD1306 driver.

**Comment by salkinium** [comment]
You can use the RPI GPIO drivers (`:platform:gpio`) in our Software I2C Bitbang driver (`:platform:i2c.bitbang`), which is feature complete. It may not be very efficient and it'll be blocking, but it should work fine.

**Comment by salkinium** [comment]
Although I'm not sure how to set OpenDrain mode with WiringPi, but @henrikssn used it with our I2C and SPI Bitbang drivers I believe?
https://github.com/modm-io/modm/blob/develop/src/modm/platform/gpio/rpi/pin.hpp

**Comment by henrikssn** [comment]
I only tried the SPI bitbang interface with RPi, it worked fine at a modest 50kHz clock.
